### DISCUSSION TITLE
Google Analytics 4  GA4

### DISCUSSION THREAD

**Discussion by seandawes** [solidusio/solidus#4621] [New Features or Ideas]
Google Analytics is becoming end of life and Google will be moving everyone to GA4. Stores are or will be needing to migrate very soon to GA4 and will lose all historical data from their analytics account potentially within a year.

I would deem this as a priority for the platform 

Additional Info
https://developers.google.com/analytics/devguides/collection/ga4/ecommerce?client_type=gtag

Currently Solidus docs outline old GA support
https://guides.solidus.io/developers/integrations/google-analytics

**Comment by fthobe** [comment]
Hey @jarednorman , 

Is this core or custom?

It has implications also for https://github.com/solidusio-contrib/solidus_multi_domain as variables should be stored there.

**Reply by jarednorman** [reply]
I would like the starter frontend to ship with this as an option.

**Reply by fthobe** [reply]
Do you want to store the variables in the backend or storefront configuration? I would suggest that there is a set of per storefront options together with the local. It makes more sense to keep the major variables together and it would also work together well with #5897. It work also require a separate taxon for brands to identify them. How do you view cookie consent?

**Reply by jarednorman** [reply]
Powering this feature with metadata is a good idea. I like that. I don't think we should make too many assumptions about how people want to keep track of brand. Anything we do there should be customizable, as some stores track that in custom ways (meaning not with taxons, though that is a sensible approach.)

Cookie consent is outside the scope of what I think should be included. Most stores use third-party consent tools and I'm skeptical building our own solution is worth it, especially since these tools are relatively plug-and-play.

**Reply by fthobe** [reply]
Jared, the requirements to have structured data and google analytics ecommerce running decently are just not there, it would be a mess regarding the difference between taxons and brands, even mono brand companies mark up their products with a brand: 
- [Example H&M markup with Brand](https://www2.hm.com/it_it/productpage.1245626001.html)
- [Example Adidas markup with brand](https://www.adidas.it/giacca-imbottita-helionic-hooded/HN5641.html)

I could understand you if we were talking about a remote field in fifteens facet of some self hosting analytics tool, but having optional fields to fetch for basic analytics 

I mean I fully understand your stance, but in my understanding also looking at commit history, I had this impression: 
- additional business logic goes into the extensions 
- basic information go into core

It just doesn't make sense to pull up extensions left and right for 10 fields that are universal and don't break anything. I just feel neither of us dominates what's the standard here, but Google / Facebook does. 

With 10 fields split up between channel, products and categories this could be an SEO and analytics machine.

Look, i'd sit down and implement the whole analytics suite and structured data, we have to do that anyway for our installation, but the 10 fields needed across the three major solutions: 
- Google Structured Data
- Google Merchant with Structured Data 
- Google Analytics
with all due respect, it seems like a baseline to me. 

Now getting banned? :)

**Reply by fthobe** [reply]
Just for ref #5905

**Reply by jarednorman** [reply]
I didn't say we couldn't use a "Brand" taxonomy to track brands. I just said that we need to make that configurable for all the stores that store that information differently. We work with multiple stores that have first-class support for a brand model that they've built out that tracks information relevant for them. While I change to the starter frontend won't directly affect them, we know that's a use case that we should be able to gracefully handle (and it's not hard to).

**Reply by fthobe** [reply]
Thank you for the detailed reply. Do you have a preferred strategy? 

I would just add a field brand that picks from the existing taxonomies, but I am more than open for suggestions!

**Reply by jarednorman** [reply]
I'm not sure what you mean "field brand that picks from existing taxonomies".

My suggestion would be to add a `Spree::Product#brand` method that looks up what taxon the product is in with in a "Brand" taxonomy, if one exists. Something like that.

**Reply by fthobe** [reply]
Ok, Google foresees a one to one relationship. Please elaborate on what you feel is an integration you would support and would create the needed outcome keeping in mind that we don't have the experience you have on this platform.

**Reply by fthobe** [reply]
> I'm not sure what you mean "field brand that picks from existing taxonomies".

A simple field similar to the current one for taxons with one allowed value instead of many.

**Reply by fthobe** [reply]
> I'm not sure what you mean "field brand that picks from existing taxonomies".
> 
> 
> 
> My suggestion would be to add a `Spree::Product#brand` method that looks up what taxon the product is in with in a "Brand" taxonomy, if one exists. Something like that.

I struggle with taxon#brand as it opens doors to varying results and multiple brands (I also find replicating the current taxon creator with a limitation to one item more charming for the obvious reason that it creates much less work.

**Reply by kennyadsl** [reply]
I don't think @jarednorman is suggesting to have multiple brands, but connect the brand to one category/taxon. 
My suggestion here is to make it easily configurable, so anyone can easily replace their own logic easily. In Pseudo-Code it is something like:

```ruby
# app/models/spree/product.rb

class Spree::Product
  def brand
    Spree::Config.brand_selector_class.new(product).call
  end
end

# app/models/spree/taxon_brand_selector.rb

class Spree::TaxonBrandSelector
  attr_reader :product

  def initialize(product)
    @product = product
  end

  def call
    product.taxons.joins(:taxonomy).where(spree_taxonomies: { name: "Brands" }).first
  end
end

# app/models/app_configuration.rb

class_name_attribute :brand_selector_class, default: 'Spree::TaxonBrandSelector'
```



This way, we have a simple, default primitive in core and anyone can easily extend this by configuring their own class:

```ruby
Spree::Config.brand_selector_class = MySuperComplexBrandSelector
```

Sounds good?

**Reply by fthobe** [reply]
Works for me

**Reply by jarednorman** [reply]
Indeed that is what I meant.

**Reply by fthobe** [reply]
@kennyadsl @jarednorman @shahmayur001 

Maybe a stupid question: 
This means we need to cycle all product taxons upon page view or am I wrong?

I am sorry, the more I think about it the more shady fix this seems to me:

- We need always to cycle all product taxons 
- we need to implement additional filters on views 
- we need to apply filters on api calls 
- we have a one to many verification to be made to assure that only one brand is taken (how do we decide the order of this) 
- if for heaven's sake you want that in an extension we can use the sales price extension and mutate it into an SEO fields extension also adding the missing fields for brands

**Reply by kennyadsl** [reply]
@fthobe `product.taxons.joins(:taxonomy).where(spree_taxonomies: { name: "Brands" }).first` this is a single query. No need to cycle all the product taxons.

**Reply by fthobe** [reply]
@kennyadsl Well, on taxons that would n calls for n products, or not?

**Reply by kennyadsl** [reply]
It depends on the context in which you need this. It could generate N+1 if you need all the products with their brand, but that's valid for any query. In that case, you can use Rails to [eager load associations](https://guides.rubyonrails.org/active_record_querying.html#eager-loading-associations).

**Reply by fthobe** [reply]
Guys, I don't know, but it seems like blowing a lot of smoke to avoid one column on a product.

@jarednorman excuse my ignorance, I am not a ruby guy and my capacities here are limited. 

As far as I understand we need to
- [ ] add a column to taxons
- [ ] make a query on every item on a category page to get the taxon and make a separate query for every product page

I understand your point, I believe backwards compatibility is important. But do you really think we can't find two names to make this backward compatible and add this to the core in a way that makes more sense (hence the product table with a reference to one taxon) without hell breaking loose?

**Reply by kennyadsl** [reply]
@fthobe I think it would be complex anyway. The brand as a string in a new product column is not enough most of the times. Don't we need users to click on the brand page that lists all the brand's product? Don't we need a description, SEO meta fields and a slug? 

By the way, why do we need to render the brand for each product on the category page?

**Reply by fthobe** [reply]
Generally improve CTR on category pages and push Google to accept a certain page hyrachy. 

It's valuable on the category pages as click through improvement also for filtering, imagine you want to see:
- all pants by brand 
- all phones by brand 

Basically everywhere where you don't have mono brand can be a factor also for filtering and facets. 

Substantially you have the four ways to do it: 

1. you are mono brand and ommit any brand information
2. you are mono brand and add the brand name and optional values (eg link, description) in the schema, it is worth saying that the brand can also be attached to the end of every title and does need to be stated twice and it's omitted on category pages.

Example: [H&M Pants](https://www2.hm.com/it_it/productpage.1260435001.html)
Title: `<title>Pantaloni sfrangiati - Nero - DONNA | H&amp;M IT</title>`
H1: `<h1 class="fa226d af6753 d582fb">Pantaloni sfrangiati</h1>`
Schema.org
```
"brand": {
      "@type": "Brand",
      "name": "H&M"
    },
```

both those cases can be covered with an if statement. 

3. you are multi brand and add the brand concatenating it with product title and separate it out in the schema additionally (eg asos) and keep the brand visible on the category page.

Example: [Reclaimed Vintage Jacket on ASOS](https://www.asos.com/reclaimed-vintage/reclaimed-vintage-branded-puffer-jacket/prd/206797685)
Title Tag `<title>Reclaimed Vintage branded puffer jacket  | ASOS</title>`
Example H1 `<h1 class="jcdpl">Reclaimed Vintage branded puffer jacket </h1>`
Schema.org
```
"brand":{
       "@type":"Brand",
       "name":"Reclaimed Vintage"
}
```

4. The most complete approach: You separate everything and concatenate where needed (eg Zalando)
Example: [Pier One 5 Pack T-Shirt on Zalando](https://www.zalando.it/pier-one-5-pack-maglietta-a-manica-lunga-bluewhitebeigeolive-pi922o0rb-a12.html) and keep the brand visible on the category page.

Title Tag: `<title>Pier One 5 PACK - Maglietta a manica lunga - white/beige/olive/bianco - Zalando.it</title>`

Example H1: 
```
<h1 class="sDq_FX _2kjxJ6 FxZV-M HlZ_Tf" data-trckng-component="YnTq0OKdconkiJJtDggo6">
    <div class="XKeLfu lm1Id5">
        <a tabindex="0" class="_ZDS_REF_SCOPE_ CDD5jc l9TUOd CKDt_l LyRfpJ JT3_zV _5Yd-hZ _5Yd-hZ" href="https://www.zalando.it/pier-one/" title="Pier One">
            <span class="_ZDS_REF_SCOPE_ _5FHGm_">
                <span class="z2N-Fg yOtBvf FxZV-M HlZ_Tf _5Yd-hZ">Pier One</span>
            </span>
        </a>
    </div>
        <span class="AGx8tI yOtBvf FxZV-M HlZ_Tf wYGQO3">
            <span class="EKabf7 R_QwOV">5 PACK - Maglietta a manica lunga</span>
        </span>
</h1>
```

Schema.org Extract:
```
"brand":{
         "@type":"Brand",
         "name":"Pier One"},
```
